bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java

```
@@ -0,0 +1,333 @@
+        synchronized (providedObjectsHolder) {
+            return providedObjectsHolder.values();
+        }
+    }
+
+    @Override
+    public void removeProviderChangeListener(ProviderChangeListener<E> listener) {
+        synchronized (listeners) {
+            listeners.remove(listener);
+        }
+    }
+
+    protected void addParser(Parser<E> parser, Map properties) {
+        String parserType = (String) properties.get(Parser.FORMAT);
+        parserType = parserType == null ? Parser.FORMAT_JSON : parserType;
+        parsers.put(parserType, parser);
+        List<URL> value = urls.get(parserType);
+        if (value != null && !value.isEmpty()) {
+            for (URL url : value) {
+                importFile(parserType, url);
+            }
+        }
+    }
+
+    protected void removeParser(Parser<E> parser, Map properties) {
+        String parserType = (String) properties.get(Parser.FORMAT);
+        parserType = parserType == null ? Parser.FORMAT_JSON : parserType;
+        parsers.remove(parserType);
+    }
+
+    protected void importResources(File file) {
+        if (file.exists()) {
+            File[] files = file.listFiles();
+            if (files != null) {
+                for (File f : files) {
+                    importResources(f);
+                }
+            } else {
+                try {
+                    URL url = file.toURI().toURL();
+                    String parserType = getParserType(url);
+                    importFile(parserType, url);
+                } catch (MalformedURLException e) {
+                    // should not happen
+                }
+            }
+        }
+    }
+
+    protected void remove(File path) {
+        try {
+            URL url = path.toURI().toURL();
+            synchronized (providerPortfolio) {
+                removeElements(providerPortfolio.remove(url));
+            }
+        } catch (MalformedURLException e) {
+            // should not happen
```

If it should not happen, we can write a log message if it happen and not silently ignore the exception.